### PR TITLE
After tab1 finish authentication, make sure that rootAuthenticationSe…

### LINKED ISSUE #23880: After tab1 finish authentication, make sure that authenticationSession is expired shortly
### Description

This is follow-up on https://github.com/keycloak/keycloak/issues/12406 .

Before #12406, when user had opened more browser tabs with login screen, then after successful authentication in tab1, the whole `RootAuthenticationSessionModel` was removed on the server from the infinispan cache. This caused that other browser tabs were not able to finish authentication. And hence needed to display "You are already logged-in" screen when user clicked on them back.

With the changes in #12406, the whole `RootAuthenticationSessionModel` is not removed anymore. Instead only single `AuthenticationSessionModel` of the current browser tab is removed and other browser tabs has a way to automatically finish authentication (which is done automatically by javascript script `authChecker.js`, which monitors if some other browser tab authenticated). Hence the `RootAuthenticationSessionModel` is usually removed as well in next turn due the other browser tabs finish authentication automatically.

The problem is, that javascript (triggered in `authChecker.js`) may not be always triggered in all browser tabs and hence some browser remain unauthenticated. This is the case for instance when some browser tab was redirected to 3rd party identity provider etc.

Due to this, it will be good to make sure that `RootAuthenticationSessionModel` is removed shortly after authentication is finished in tab1 to make sure that there is not unused authentication sessions in memory. As this could result in growing count of items in the `authenticationSessions` infinispan cache.

This should be good compromise between memory consumptions (as rootAuthenticationSession is removed at least 1 minute after successful authentication) and usability (other browser tabs have opportunity to finish authentication due the javascript).

### Possible solution

After authentication in tab1, make sure that `RootAuthenticationSessionModel` is scheduled for expiration in short interval. This can be likely realm `accessCodeLifespan` interval, which is 1 minute by default. This can be triggered in `AuthenticationSessionManager.updateAuthenticationSessionAfterSuccessfulAuthentication` .

### PR DESCRIPTION
**Pull request by mposolda**
…ssion is expired shortly

closes #23880

### CODE REVIEW

**Comment by mposolda** [comment]
@keycloak/store-maintainers Can I please ask someone from the store team to review this? Please see the description for more details. Feel free to ping me if something is unclear.

File: services/src/main/java/org/keycloak/services/managers/AuthenticationSessionManager.java [hunk 1]
```diff
@@ -262,15 +263,22 @@ public boolean removeTabIdInAuthenticationSession(RealmModel realm, Authenticati
...
-            log.tracef("Removed authentication session of root session '%s' with tabId '%s'. But there are remaining tabs in the root session", rootAuthSession.getId(), authSession.getTabId());
+            // 1 minute by default. Same timeout, which is used for client to complete "authorization code" flow
+            // Very short timeout should be OK as when this cookie is set, other existing browser tabs are supposed to be refreshed immediately by JS script authChecker.js
+            // and login user automatically. No need to have authenticationSession and cookie living any longer
+            int authSessionExpiresIn = realm.getAccessCodeLifespan();
 
-            AuthenticationStateCookie.generateAndSetCookie(session, realm, rootAuthSession);
+            // Set timestamp to the past to make sure that authSession is scheduled for expiration in "authSessionExpiresIn" seconds
+            int authSessionExpirationTime = Time.currentTime() - SessionExpiration.getAuthSessionLifespan(realm) + authSessionExpiresIn;
+            rootAuthSession.setTimestamp(authSessionExpirationTime);
```
Comment (rmartinc) [inline services/src/main/java/org/keycloak/services/managers/AuthenticationSessionManager.java:277]: It took me some time understand that we are moving the timestamp to the past to expire it in `authSessionExpiresIn`. As the `RootAuthenticationSessionAdapter` is using the timestamp to calculate the expiration. Just a comment for other reviewers. :smile:

**Review by rmartinc — APPROVED** [review]
LGTM!

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/LoginTest.java [hunk 2]
```diff
@@ -894,6 +893,28 @@ public void openLoginFormAfterExpiredCode() throws Exception {
...
+
+        loginPage.login("test-user@localhost", "password");
+        appPage.assertCurrent();
+
+        // authentication session should still exists with remaining browser tab
+        Assert.assertEquals((Integer) 1, getTestingClient().testing().getAuthenticationSessionTabsCount("test", authSessionId));
+
+        // authentication session should be expired after 1 minute
+        setTimeOffset(300);
+        Assert.assertEquals((Integer) (-1), getTestingClient().testing().getAuthenticationSessionTabsCount("test", authSessionId));
```
Comment (thomasdarimont) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/LoginTest.java:914]: Looks a but odd to see -1 here if the last remaining auth session was removed. I'd expect this to be 0.

Comment (martin-kanis) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/LoginTest.java:914]: Yeah it looks weird, given the method name. Would it be sufficient return also 0 when there in no root auth session?

File: services/src/main/java/org/keycloak/services/managers/AuthenticationSessionManager.java (hunk 1 above)
Comment (martin-kanis) [reply services/src/main/java/org/keycloak/services/managers/AuthenticationSessionManager.java:277]: I think, I don't like this kind of hacks, but given the fact we don't have expiration field for expirable entities as we had in new store, I don't see any other option. Just a thought for possible future refactoring.

Comment (mposolda) [reply services/src/main/java/org/keycloak/services/managers/AuthenticationSessionManager.java:277]: +1, expiration field would be nice in the future.

**Review by martin-kanis — APPROVED** [review]
Thanks @mposolda. LGTM

**Comment by mposolda** [comment]
@rmartinc @martin-kanis @thomasdarimont Thanks everyone for the review!

**Pull request merged**